src/pybind/mgr/cephadm/upgrade.py

```diff
@@ -1176,6 +1199,27 @@ def _do_upgrade(self):
 
         image_settings = self.get_distinct_container_image_settings()
 
+        # unset mon flags for upgrade
+        if self.upgrade_state.mon_flags:
+            try:
+                flags_toggled = list()
+                for flag in self.upgrade_state.mon_flags:
+                    if self.mgr.is_global_mon_flag_set(flag):
+                        self.mgr.unset_global_mon_flag(flag)
+                        flags_toggled.append(flag)
+                # set mon_flags to the list of flags actually unset before upgrade.
+                # this is to ensure we only re-set these flags at end of upgrade 
+                # and don't mistakenly set any flag that wasn't set before upgrade. 
+                self.upgrade_state.mon_flags = flags_toggled
```

Comment:
there might be a slight issue with doing this here. In a single upgrade I'd expect `_do_upgrade` to run a number of times. It looks like we're building the list of flags stored in the `mon_flags` attribute of the upgrade state based on which flags we flip here. If we run through this multiple times and the first iteration unsets all the flags, I'd expect the follow up iterations to not have anything to set and we set the `mon_flags` attribute to an empty list. It looks like we're then using that attribute to set the flags again at the end. We might have to do this in the mgr module as part of the `upgrade_start` function instead and then set the mon flag attributes once in the initial upgrade state object created there so that we don't lose what we set during the upgrade